pallets/xvm/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+pub use weights::WeightInfo;
+
+#[cfg(test)]
+mod mock;
+
+pub use pallet::*;
+
+pub type WeightInfoOf<T> = <T as Config>::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_contracts::Config {
+        /// Mapping from `Account` to `H160`.
+        type AccountMapping: AccountMapping<Self::AccountId>;
+
+        /// Mapping from Ethereum gas to Substrate weight.
+        type GasWeightMapping: GasWeightMapping;
+
+        /// `CheckedEthereumTransact` implementation.
+        type EthereumTransact: CheckedEthereumTransact;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+}
+
+impl<T: Config> XvmCall<T::AccountId> for Pallet<T> {
+    fn call(
+        context: Context,
+        vm_id: VmId,
+        source: T::AccountId,
+        target: Vec<u8>,
+        input: Vec<u8>,
+    ) -> CallResult {
+        Pallet::<T>::do_call(context, vm_id, source, target, input, false)
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    fn do_call(
+        context: Context,
+        vm_id: VmId,
+        source: T::AccountId,
+        target: Vec<u8>,
+        input: Vec<u8>,
+        skip_execution: bool,
+    ) -> CallResult {
+        ensure!(
+            context.source_vm_id != vm_id,
+            CallErrorWithWeight {
+                error: CallError::SameVmCallNotAllowed,
+                used_weight: match vm_id {
+                    VmId::Evm => WeightInfoOf::<T>::evm_call_overheads(),
+                    VmId::Wasm => WeightInfoOf::<T>::wasm_call_overheads(),
+                },
+            }
+        );
+
+        match vm_id {
+            VmId::Evm => Pallet::<T>::evm_call(context, source, target, input, skip_execution),
+            VmId::Wasm => Pallet::<T>::wasm_call(context, source, target, input, skip_execution),
+        }
+    }
+
+    fn evm_call(
+        context: Context,
+        source: T::AccountId,
+        target: Vec<u8>,
+        input: Vec<u8>,
+        skip_execution: bool,
+    ) -> CallResult {
+        log::trace!(
+            target: "xvm::evm_call",
+            "Calling EVM: {:?} {:?}, {:?}, {:?}",
+            context, source, target, input,
+        );
+
+        let target_decoded =
+            Decode::decode(&mut target.as_ref()).map_err(|_| CallErrorWithWeight {
+                error: CallError::InvalidTarget,
+                used_weight: WeightInfoOf::<T>::evm_call_overheads(),
+            })?;
+        let bounded_input = BoundedVec::<u8, ConstU32<MAX_ETHEREUM_TX_INPUT_SIZE>>::try_from(input)
+            .map_err(|_| CallErrorWithWeight {
+                error: CallError::InputTooLarge,
+                used_weight: WeightInfoOf::<T>::evm_call_overheads(),
+            })?;
+
+        if skip_execution {
```

Comment:
Would it be more appropriate to move this code on L166?

Not sure how compiler would optimize it TBH, but even the minor operations before `xvm_transact` invocation will add something little to the overall weight.

Reply:
Good catch. It was right before the `xvm_transact` execution previously but was moved in this [commit](https://github.com/AstarNetwork/Astar/pull/980/commits/34a3942e12c671c6a5c6792187605484e41e4f9a). Can't remember what I was thinking. - -

Reply:
Just noted `wasm_call` has a similar problem. Updated both of them and will re-do benchmarks.